br/pkg/backup/prepare_snap/env.go

```diff
@@ -0,0 +1,171 @@
+)
+
+type Env interface {
+	ConnectToStore(ctx context.Context, storeID uint64) (PrepareClient, error)
+	GetAllLiveStores(ctx context.Context) ([]*metapb.Store, error)
+
+	LoadRegionsInKeyRange(ctx context.Context, startKey, endKey []byte) (regions []Region, err error)
+}
+
+type PrepareClient interface {
+	Send(*brpb.PrepareSnapshotBackupRequest) error
+	Recv() (*brpb.PrepareSnapshotBackupResponse, error)
+}
+
+type SplitRequestClient struct {
+	PrepareClient
+	MaxRequestSize int
+}
+
+func (s SplitRequestClient) Send(req *brpb.PrepareSnapshotBackupRequest) error {
+	// Try best to keeping the request untouched.
+	rs := req.Regions
+	if req.Ty == brpb.PrepareSnapshotBackupRequestType_WaitApply && req.Size() > s.MaxRequestSize {
+		findSplitIndex := func() int {
+			if len(rs) == 0 {
+				return -1
+			}
+
+			// Select at least one request.
+			// So we won't get sutck if there were a really huge (!) request.
+			collected := 0
+			lastI := 1
+			for i := 2; i < len(rs) && collected+rs[i].Size() < s.MaxRequestSize; i++ {
+				lastI = i
+				collected += rs[i].Size()
+			}
+			return lastI
+		}
+		for splitIdx := findSplitIndex(); splitIdx > 0; splitIdx = findSplitIndex() {
+			split := &brpb.PrepareSnapshotBackupRequest{
+				Ty:      brpb.PrepareSnapshotBackupRequestType_WaitApply,
+				Regions: rs[:splitIdx],
+			}
+			rs = rs[splitIdx:]
+			if err := s.PrepareClient.Send(split); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	return s.PrepareClient.Send(req)
+}
+
+type Region interface {
+	GetMeta() *metapb.Region
+	GetLeaderStoreID() uint64
+}
+
+type CliEnv struct {
+	Cache *tikv.RegionCache
+	Mgr   *utils.StoreManager
+}
+
+func (c CliEnv) GetAllLiveStores(ctx context.Context) ([]*metapb.Store, error) {
+	stores, err := c.Cache.PDClient().GetAllStores(ctx, pd.WithExcludeTombstone())
+	if err != nil {
+		return nil, err
+	}
+	withoutTiFlash := slices.DeleteFunc(stores, engine.IsTiFlash)
+	return withoutTiFlash, err
+}
+
+func (c CliEnv) ConnectToStore(ctx context.Context, storeID uint64) (PrepareClient, error) {
+	var cli brpb.Backup_PrepareSnapshotBackupClient
+	err := c.Mgr.TryWithConn(ctx, storeID, func(cc *grpc.ClientConn) error {
+		bcli := brpb.NewBackupClient(cc)
+		c, err := bcli.PrepareSnapshotBackup(ctx)
+		if err != nil {
+			return errors.Annotatef(err, "failed to create prepare backup stream")
+		}
+		cli = c
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return cli, nil
+}
+
+func (c CliEnv) LoadRegionsInKeyRange(ctx context.Context, startKey []byte, endKey []byte) (regions []Region, err error) {
+	bo := tikv.NewBackoffer(ctx, regionCacheMaxBackoffMs)
+	if len(endKey) == 0 {
+		// This is encoded [0xff; 8].
+		// Workaround for https://github.com/tikv/client-go/issues/1051.
+		endKey = []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
+	}
+	rs, err := c.Cache.LoadRegionsInKeyRange(bo, startKey, endKey)
+	if err != nil {
+		return nil, err
+	}
+	rrs := make([]Region, 0, len(rs))
```

Comment:
why need we make a local copy?

Reply:
Because the type doesn't match. The interface `RegionCache` returns is differ from the `Region` type we are returning.
Anyway this will be a shallow copy and won't be very expensive.